Add upgrade documentation (#1798)
graphql-dotnet/graphql-dotnet: docs2/site/docs/guides/migration3.md

# Migrating from v2.x to v3.x

## New Features

### Support for additional scalar types and conversions built-in

These .Net types now are automatically mapped to corresponding built-in custom scalar types:
* Byte
* SByte
* Short
* UShort
* UInt
* ULong
* Guid (maps to `IdGraphType` by default; also supports `GuidGraphType`)
* BigInt

There is also support for converting byte arrays to base-64 encoded strings and vice versa. See
the `ValueConverter` class for more details.

See [Schema Types](https://graphql-dotnet.github.io/docs/getting-started/schema-types) for more details.

### Other new features

* Based on .Net Standard 2.0, supporting .Net Core applications
* Supports scoped services (see below under [Dependency Injection](#dependency-injection))

> * Supports the `System.Text.Json` library for JSON serialization, in addition to the Newtonsoft.Json library

Reviewer: ```suggestion
* Supports the `System.Text.Json` package for JSON serialization, in addition to the `Newtonsoft.Json` package
```

> * Data loaders work with serial execution strategies and can be chained together

Reviewer: Does it require #1511 ?

Author: Yes

* Field builders can take an optional configuration action parameter
* Support for auto-registering input object graph types via `AutoRegisteringInputObjectGraphType`
* Added codes to `ExecutionError`s
* Document processing exceptions can be logged or modified (see below under [Exception Handling](#exception-handling))
* Enhanced validation of graphs built-in
* Supports filtering of schema introspection requests - see details [here](https://github.com/graphql-dotnet/graphql-dotnet/pull/1179)
* Supports federated schemas - see details [here](https://github.com/graphql-dotnet/graphql-dotnet/pull/1233)
* Supports schema 'description' property - see details [here](https://github.com/graphql-dotnet/graphql-dotnet/pull/1613)
* Supports comment nodes - see details [here](https://github.com/graphql-dotnet/graphql-dotnet/pull/1617)
* Supports result 'extensions' - see details [here](https://github.com/graphql-dotnet/graphql-dotnet/pull/1611)
* Ability to limit maximum number of asynchronous field resolvers executing simultaneously - see details [here](https://github.com/graphql-dotnet/graphql-dotnet/issues/1239)

## Breaking Changes

> ### .Net compatibility

Reviewer: ```suggestion
### .NET compatibility
```

Please change .Net to .NET everywhere in this file.

> This project now requires the .Net Standard 2.0 framework, and breaks compatibility with applications based on

Reviewer: ```suggestion
This project now requires the .NET Standard 2.0 and breaks compatibility with applications based on
```

Using `framework` along with .NET Standard can confuse someone.

> .Net Framework 4.6 and earlier.  See https://docs.microsoft.com/en-us/dotnet/standard/net-standard for a list of

Reviewer: ```suggestion
.NET Framework 4.6 and earlier. See https://docs.microsoft.com/en-us/dotnet/standard/net-standard for a list of
```

frameworks that support .Net Standard 2.0.

### Dependency Injection

The previous `IDependencyResolver` interface and `FuncDependencyResolver` class have been replaced by the
.Net Standard `IServiceProvider` interface and the new `FuncServiceProvider` class.

```csharp
//public Schema(IDependencyResolver dependencyResolver)
public Schema(IServiceProvider serviceProvider)
{
  ...
}
```

Also, the `Schema.DependencyResolver` property has been removed and not replaced. If you need to access the service provider
from your graphs, you can include `IServiceProvider` in the constructor of the graph type. Your DI container will pass

> a reference to the service provider.  For singleton schemas, this will be the root service provider, from which you can

Reviewer: ```suggestion
a reference to the service provider. For singleton schemas, this will be the root service provider, from which you can
```

obtain other singleton or transient services. For scoped schemas with scoped graph types, this will be the service provider
for the current executing scope. Casting `Schema` to `IServiceProvider` is also possible, but not recommended, and will
yield similar results.

If you wish to access a scoped service from within a resolver and want to use a singleton schema (as is recommended), you
can pass a scoped service provider to `ExecutionOptions.RequestServices`, which can then be used to resolve scoped

> services. For Asp.Net Core projects, you can set this to equal `HttpContext.RequestServices`. Be aware that if you

Reviewer: ```suggestion
services. For ASP.NET Core projects, you can set this equal to `HttpContext.RequestServices`. Be aware that if you
```

    (
        context.Request.Body,
        new JsonSerializerOptions { PropertyNameCaseInsensitive = true }
    );

    var executer = new DocumentExecuter();
    var result = await executer.ExecuteAsync(options =>
    {
        options.Schema = schema;
        options.Query = request.Query;
        options.OperationName = request.OperationName;
        options.Inputs = request.Variables.ToInputs();
    });

    context.Response.ContentType = "application/json";
    context.Response.StatusCode = 200; // OK

    var writer = new GraphQL.SystemTextJson.DocumentWriter();
    await writer.WriteAsync(context.Response.Body, result);
}

public class GraphQLRequest
{
    public string OperationName { get; set; }

    public string Query { get; set; }

    [JsonConverter(typeof(GraphQL.SystemTextJson.ObjectDictionaryConverter))]
    public Dictionary<string, object> Variables { get; set; }
}
```

This is just a simplified example of the changes necessary. Note that typically the `DocumentExecuter` and `DocumentWriter`
are registered as singletons within the dependency injection container, as they can safely be shared between requests.

> If you continue to use the `Newtonsoft.Json` converter, please note that Asp.Net Core 3.0 disallows synchronous IO by default,

Reviewer: ```suggestion
If you continue to use the `Newtonsoft.Json` converter, please note that ASP.NET Core 3.0 disallows synchronous IO by default,
```

> which is required by the converter.  You will need to make a change in the `ConfigureServices` section of `Startup.cs` as

Reviewer: ```suggestion
which is required by the converter. You will need to make a change in the `ConfigureServices` section of `Startup.cs` as
```

follows:

```csharp
// kestrel
services.Configure<KestrelServerOptions>(options =>
{
    options.AllowSynchronousIO = true;
});

// IIS
services.Configure<IISServerOptions>(options =>
{
    options.AllowSynchronousIO = true;
});
```

Please note that if you use a `NameConverter` other than the default `CamelCaseNameConverter`, you may need to configure
your json serializer also to not convert object properties to camel case. For example, with the `System.Text.Json` converter,
you need to set `JsonSerializerOptions.PropertyNamingPolicy = null;` as follows:

```csharp
var writer = new GraphQL.SystemTextJson.DocumentWriter(options => {
    options.PropertyNamingPolicy = null;
});
```

### UserContext

The definition of the `UserContext` object throughout the library has changed to a `IDictionary<string, object>`.
Custom user context classes will need to inherit from `Dictionary<string, object>` or otherwise support the interface.

```csharp
//class MyContext
class MyContext : Dictionary<string, object>
{
    public DbContext MyDbContext { get; set; }
}
```

> 

Reviewer: ```suggestion

If you are going to use GraphQL.NET in ASP.NET Core application, then you may be interested in the [server](https://github.com/graphql-dotnet/server) project. It already implements all the necessary integration with GraphQL.NET via the custom ASP.NET Core middleware.
```

### Document Listeners

The `DocumentExecutionListenerBase<T>` class and `IDocumentExecutionListener<T>` interface have been removed;
please implement the `IDocumentExecutionListener` interface when creating a custom document listener. You
can also inherit from the `DocumentExecutionListenerBase` class to provide default implementations of events.

The methods definitions have also changed from passing `userContext` and `cancellationToken` parameters to a

> single parameter `context` of type `IExecutionContext`.  The context has properties for accessing the user context,

Reviewer: ```suggestion
single parameter `context` of type `IExecutionContext`. The context has properties for accessing the user context,
```

Author: So dumb that html doesn't support two spaces.  Not that there's anything that can be done about it considering the design of html.  It just aggravates me to no end to see improperly spaced paragraphs.

cancellation token, metrics, execution errors, and other information about the executing request.

```csharp
//class MyListener : DocumentExecutionListenerBase<MyContext>
class MyListener : DocumentExecutionListenerBase
{
    //public virtual Task AfterValidationAsync(MyContext userContext, IValidationResult validationResult, CancellationToken token)
    public Task AfterValidationAsync(IExecutionContext context, IValidationResult validationResult)
    {
        var myContext = (MyContext)context.userContext;

        // log validation error

        return Task.CompletedTask;
    }

    ...
}
```

### IResolveFieldContext and IResolveConnectionContext

> Field resolver methods now are passed a reference to an `IResolveFieldContext` interface, rather than a

Reviewer: ```suggestion
Field resolver methods now pass a reference to an `IResolveFieldContext` interface, rather than a
```

> `ResolveFieldContext` class.  Inline lambda functions are typically unaffected, but if you define your

Reviewer: ```suggestion
`ResolveFieldContext` class. Inline lambda functions are typically unaffected, but if you define your
```

the signature for the `Resolve` method to accept `IResolveFieldContext`.

```csharp
//class MyMiddleware
class MyMiddleware : IFieldMiddleware
{
    //public async Task<object> Resolve(ResolveFieldContext context, FieldMiddlewareDelegate next)
    public async Task<object> Resolve(IResolveFieldContext context, FieldMiddlewareDelegate next)
    {
        // your code here
        var ret = await next(context);
        // your code here
        return ret;
    }
}
```

Please note that the delegate definition for `FieldMiddlewareDelegate` has changed as follows:

```csharp
// version 2.4.0
public delegate Task<object> FieldMiddlewareDelegate(ResolveFieldContext context);
// version 3.0
public delegate Task<object> FieldMiddlewareDelegate(IResolveFieldContext context);
```

You also must ensure that your schema implements `IServiceProvider`. This is handled
automatically if you inherit from `Schema`.

See [Field Middleware](https://graphql-dotnet.github.io/docs/getting-started/field-middleware) for more details,
including guidelines and restrictions on service lifetimes of middleware registered through your DI framework.

### Data Loaders

> Data loaders now return an `IDataLoaderResult<T>` rather than a `Task<T>`.  Field resolver signatures may need to change

Reviewer: ```suggestion
Data loaders now return an `IDataLoaderResult<T>` rather than a `Task<T>`. Field resolver signatures may need to change
```

> as a result.  Lambda functions passed to field builders' `ResolveAsync` method should not need to change.

Reviewer: ```suggestion
as a result. Lambda functions passed to field builders' `ResolveAsync` method should not need to change.
```

    }
}
```

If you need to process the data loader result before it is returned, additional refactoring will need to be done.
The data loader also now supports chained data loaders, and asynchronous code prior to queuing the data loader. See
[Data loader documentation](https://graphql-dotnet.github.io/docs/getting-started/dataloader) for more details.

### ExecutionStrategy changes

If you utilize data loaders along with a custom implementation of `IExecutionStrategy` (typically inheriting
from `ExecutionStrategy`), you must change the implementation to monitor for `IDataLoaderResult` returned
values, and execute `GetResultAsync` at the appropriate time to retrieve the actual value asynchronously.
If the field resolver returns a `Task<IDataLoaderResult>`, the execution strategy should start the task as
usual, only queuing the data loader once the `IDataLoaderResult` has been returned. Note that
`await IDataLoaderResult.GetResultAsync()` may return another `IDataLoaderResult` which must again
be queued to execute at the proper time. Please refer to the reference implementation of `ParallelExecutionStrategy`
for an example.

### Exception Handling

> Exceptions have been split into three categories: schema errors, input errors, and processing errors.  For instance,

Reviewer: ```suggestion
Exceptions have been split into three categories: schema errors, input errors, and processing errors. For instance,
```

if an invalid query was passed to the `DocumentExecuter`, it would be considered an input error, and a `SyntaxError`
would be thrown. Or if an invalid enum string was passed as a variable to a query, an `InvalidValueError` would be
thrown. All validation rules that fail their respective tests are treated as input errors.

In 2.x, most schema errors would throw an exception of the `ExecutionError` type. This has been changed; now
the thrown error is a native exception -- for instance, an `ArgumentOutOfRange` exception would be thrown when
trying to add a field to a type with the same name as one that already exists.

In 3.x, `ExecutionError` (and derived classes) are returned directly from the `DocumentExecuter` for parsing

> and validation errors (input errors). Field resovlers and middleware can also return input errors by

Reviewer: ```suggestion
and validation errors (input errors). Field resolvers and middleware can also return input errors by
```

of the `DocumentExecuter` by setting `ThrowOnUnhandledException` to `true`.

See [Error handling documentation](https://graphql-dotnet.github.io/docs/getting-started/errors) for more details,
samples, and a full list of exception types and codes.

### SubscriptionExecuter removal

The `SubscriptionExecuter` class, previously marked as obsolete, has been removed. Use the `DocumentExecuter`
in its place.

### NameConverter

The `ExecutionOptions.FieldNameConverter` property has been replaced by the `NameConverter` property, and the
corresponding class names have changed as well. Static instance members have been added to the included
name converters. Below is a sample of the required change when using `PascalCaseFieldNameConverter`:

```csharp
var executer = new DocumentExecuter();
var result = executer.ExecuteAsync(options => {

	...

  //options.FieldNameConverter = new PascalCaseFieldNameConverter();
  options.NameConverter = PascalCaseNameConverter.Instance;
});
```

If you have written a custom name converter, you must now implement `INameConverter` rather than
`IFieldNameConverter`, which has two methods, `NameForField` and `NameForArgument`, rather than

> only `NameFor`. There is also no need to check for introspection types, as the framework will

Reviewer: ```suggestion
only `NameFor`. There is also no need to check for introspection types, as the GraphQL.NET will
```

Author: ```suggestion
only `NameFor`. There is also no need to check for introspection types, as GraphQL.NET will
```

which is required by the converter. You will need to make a change in the `ConfigureServices` section of `Startup.cs` as
follows:

```csharp
// kestrel
services.Configure<KestrelServerOptions>(options =>
{
    options.AllowSynchronousIO = true;
});

// IIS
services.Configure<IISServerOptions>(options =>
{
    options.AllowSynchronousIO = true;
});
```

Please note that if you use a `NameConverter` other than the default `CamelCaseNameConverter`, you may need to configure
your json serializer also to not convert object properties to camel case. For example, with the `System.Text.Json` converter,
you need to set `JsonSerializerOptions.PropertyNamingPolicy = null;` as follows:

```csharp
var writer = new GraphQL.SystemTextJson.DocumentWriter(options => {
    options.PropertyNamingPolicy = null;
});
```

### UserContext

The definition of the `UserContext` object throughout the library has changed to a `IDictionary<string, object>`.
Custom user context classes will need to inherit from `Dictionary<string, object>` or otherwise support the interface.

```csharp
//class MyContext
class MyContext : Dictionary<string, object>
{
    public DbContext MyDbContext { get; set; }
}
```

> If you are going to use GraphQL.NET in ASP.NET Core application, then you may be interested in the [server](https://github.com/graphql-dotnet/server) project. It already implements all the necessary integration with GraphQL.NET via the custom ASP.NET Core middleware.

Author: ```suggestion
If you are going to use GraphQL.NET in an ASP.NET Core application, then you may be interested in the
[server](https://github.com/graphql-dotnet/server) project. It already implements all the necessary integration
with GraphQL.NET via the custom ASP.NET Core middleware.
```

    options.AllowSynchronousIO = true;
});

// IIS
services.Configure<IISServerOptions>(options =>
{
    options.AllowSynchronousIO = true;
});
```

Please note that if you use a `NameConverter` other than the default `CamelCaseNameConverter`, you may need to configure
your json serializer also to not convert object properties to camel case. For example, with the `System.Text.Json` converter,
you need to set `JsonSerializerOptions.PropertyNamingPolicy = null;` as follows:

```csharp
var writer = new GraphQL.SystemTextJson.DocumentWriter(options => {
    options.PropertyNamingPolicy = null;
});
```

### UserContext

The definition of the `UserContext` object throughout the library has changed to a `IDictionary<string, object>`.
Custom user context classes will need to inherit from `Dictionary<string, object>` or otherwise support the interface.

```csharp
//class MyContext
class MyContext : Dictionary<string, object>
{
    public DbContext MyDbContext { get; set; }
}
```

If you are going to use GraphQL.NET in an ASP.NET Core application, then you may be interested in the
[server](https://github.com/graphql-dotnet/server) project. It already implements all the necessary integration

> with GraphQL.NET via the custom ASP.NET Core middleware.
> ### Document Listeners

Reviewer: ```suggestion
with GraphQL.NET via the custom ASP.NET Core middleware.

### Document Listeners
```

the signature for the `Resolve` method to accept `IResolveFieldContext`.

```csharp
//class MyMiddleware
class MyMiddleware : IFieldMiddleware
{
    //public async Task<object> Resolve(ResolveFieldContext context, FieldMiddlewareDelegate next)
    public async Task<object> Resolve(IResolveFieldContext context, FieldMiddlewareDelegate next)
    {
        // your code here
        var ret = await next(context);
        // your code here
        return ret;
    }
}
```

Please note that the delegate definition for `FieldMiddlewareDelegate` has changed as follows:

```csharp
// version 2.4.0
public delegate Task<object> FieldMiddlewareDelegate(ResolveFieldContext context);
// version 3.0
public delegate Task<object> FieldMiddlewareDelegate(IResolveFieldContext context);
```

You also must ensure that your schema implements `IServiceProvider`. This is handled
automatically if you inherit from `Schema`.

See [Field Middleware](https://graphql-dotnet.github.io/docs/getting-started/field-middleware) for more details,
including guidelines and restrictions on service lifetimes of middleware registered through your DI framework.

### Data Loaders

> Data loaders now return an `IDataLoaderResult<T>` rather than a `Task<T>`. Field resolver signatures may need to change

Reviewer: I've been going to review #1511 for quite some time. I couldn't do it because I had to review many of your other PRs (which in itself is not bad, you did a lot). Time is coming to the end of August. I'm not really sure if I can make a review in time. I would not want to release the 3.0 version (or what is the same to merge #1511 in master) with this feature, which has not been reviewed, despite the fact that it has been successfully applied by someone in production. Another reason is that I also have a set of changes that I would also like to add to the 3.0 version. There are backward incompatible things - middlewares redesign, changes in SDL printing and custom scalars. The dev teams in my company are also waiting for them (I am not yet talking about bug fixes which also needs to be done a lot). I would not want to release 4.0 immediately after the 3.0 version. Let's move the deadline for 3.0 to October 1, by a month. I will have a vacation almost all September and I will be able to devote much more time to this project. Recently, I simply could not spend all day thinking about a complex PR like #1511 one. So during next month, we will be able to prepare #1511 and other features as well. I believe it is better to work a little longer without rushing and produce a better product. Another month of delay isn't such a big deal if the previous version release was almost 2 years ago.

Author: I, respectfully, disagree.  There are multiple people who use #1511 already.  We are all here to maintain the project.  Let's just merge the dataloaders and keep going.  I've had this PR up since January.  I've done all the documentation.  I've made all these other changes we've been trying to get into 3.0.  Got a ton of stuff cleared off the issue list.  Anyone who wants to review my PR has had ample time to do so.  There's no reason to wait another month.

I understand releasing a few more features with 3.0, as long as we have a **SET** deadline and very good reason for the delay.  We can just merge the data loaders and release this as another beta.  If you have a dev team waiting for the features @sungam3r then good, let's get it done and release 3.0.  Other people have dev teams waiting for stuff, though, so we can't let it drag out indefinitely.  Pick a date and stick to it, regardless of the result.  We can always release another version.

If we do simply release another 3.0 beta (with data loaders) this weekend, then if there's any changes we need to make to the data loaders, they can be done in the interim.  Breaking changes are not a big problem.  Indefinite delays are.

Reviewer 2: I fall into the camp of releasing 3.0 beta with the data loaders, 3.0, and releasing a 4.0 soon after if needed.  I can review #1511 tomorrow and propose to merge if I don't find anything major.

Reviewer: Fine. I just wanted someone else to review.
